src/main/java/org/assertj/core/api/Assertions.java

```diff
@@ -1139,12 +1139,48 @@ public static AtomicLongArrayAssert assertThat(AtomicLongArray actual) {
    *
    * @param shouldRaiseThrowable The lambda with the code that should raise the exception.
    * @return The captured exception or <code>null</code> if none was raised by the callable.
+   * @see #catchThrowableOfType(ThrowingCallable, Class)
    */
   public static Throwable catchThrowable(ThrowingCallable shouldRaiseThrowable) {
     return AssertionsForClassTypes.catchThrowable(shouldRaiseThrowable);
   }
 
   /**
+   * Allows catching an {@link Throwable} of a specific type more easily when used with Java 8 lambdas.
+   *
+   * <p>
+   * A call is made to {@link #catchThrowable(ThrowingCallable)}, and an assertion is made on the caught {@link Throwable} to ensure it is of the specified type. If
+   * the assertion passes, then the caught {@link Throwable} is cast to the correct subtype before
+   * being returned, making it convenient to perform subtype-specific assertions on the result.
+   * </p>
+   *
+   * <p>
+   * Example:
+   * </p>
+   *
+   * <pre><code class='java'>{@literal @}Test
+   * public void testException() {
+   *   // when
+   *   CustomParseException e = catchThrowableOfType(() -&gt; { throw new CustomParseException("boom!", 1, 5); },
+   *                                                 CustomParseException.class);
+   *
+   *   // then
+   *   assertThat(e).as("message").hasMessageContaining("boom");
+   *   assertThat(e.getLine()).as("line").isEqualTo(1);
+   *   assertThat(e.getColumn()).as("column").isEqualTo(5);
+   * } </code></pre>
+   *
+   * @param shouldRaiseThrowable The lambda with the code that should raise the exception.
+   * @param type The type of exception that the code is expected to raise.
+   * @return The captured exception or <code>null</code> if none was raised by the callable.
```

Comment:
I believe null is never returned as we try to cast the return to `E`

Reply:
Yes, in fact I already noticed this in my own code - the isInstanceOf() will assert if the exception is null, and the error message is not very helpful. I already fixed the error message in my local repository.

However, one concern I have with going down that path is is that of consistency between `catchThrowable()` and `catchThrowableOfType()`. The former will happy return `null` if no exception is thrown. Will it confuse developers if `catchThrowableOfType()` behaves slightly differently? I suppose it's already behaving differently because it is doing an assertion anyway, so we might as well do that. But then perhaps renaming it to something else like `assertThatBlockThrowsThrowableOfType()` would make the difference more clear?

Reply:
I was thinking the most likely use case would be for the code to raise the expected exception type but we actually can allow null and still have the return value to be cast to the expected exception.

basically it boils down to only check the instance type if not null
```java
if (t != null) {
  new ThrowableAssert(t).isInstanceOf(type);
}
return (E)t; // works if t is null
```